flow/layers/container_layer.cc

```diff
@@ -4,39 +4,74 @@
 
 #include "flutter/flow/layers/container_layer.h"
 
+#include <memory>
+
+#include "flutter/flow/layers/transform_layer.h"
+
 namespace flutter {
 
-ContainerLayer::ContainerLayer() {}
+static float ClampElevation(float elevation,
+                            float parent_elevation,
+                            float max_elevation) {
+  // TODO(mklim): Deal with bounds overflow more elegantly. We'd like to be
+  // able to have developers specify the behavior here to alternatives besides
+  // clamping, like normalization on some arbitrary curve.
+  float clamped_elevation = elevation;
+  if (max_elevation > -1 && (parent_elevation + elevation) > max_elevation) {
+    // Clamp the local z coordinate at our max bound. Take into account the
+    // parent z position here to fix clamping in cases where the child is
+    // overflowing because of its parents.
+    clamped_elevation = max_elevation - parent_elevation;
+  }
 
-ContainerLayer::~ContainerLayer() = default;
+  return clamped_elevation;
+}
 
+ContainerLayer::ContainerLayer(bool force_single_child) {
+  // Place all "child" layers under a single child if requested.
+  if (force_single_child) {
+    single_child_ = std::make_shared<TransformLayer>(SkMatrix::I());
+    single_child_->set_parent(this);
+    layers_.push_back(single_child_);
+  }
+}
 void ContainerLayer::Add(std::shared_ptr<Layer> layer) {
+  // Place all "child" layers under a single child if requested.
+  if (single_child_) {
+    single_child_->Add(std::move(layer));
+    return;
+  }
+
   layer->set_parent(this);
   layers_.push_back(std::move(layer));
 }
 
 void ContainerLayer::Preroll(PrerollContext* context, const SkMatrix& matrix) {
   TRACE_EVENT0("flutter", "ContainerLayer::Preroll");
 
-  SkRect child_paint_bounds = SkRect::MakeEmpty();
-  PrerollChildren(context, matrix, &child_paint_bounds);
-  set_paint_bounds(child_paint_bounds);
-}
+  // Track total elevation as we walk the tree, in order to deal with bounds
+  // overflow in z.
+  parent_elevation_ = context->total_elevation;
+  clamped_elevation_ = ClampElevation(elevation_, parent_elevation_,
+                                      context->frame_physical_depth);
+  context->total_elevation += clamped_elevation_;
 
-void ContainerLayer::PrerollChildren(PrerollContext* context,
-                                     const SkMatrix& child_matrix,
-                                     SkRect* child_paint_bounds) {
+  SkRect child_paint_bounds = SkRect::MakeEmpty();
   for (auto& layer : layers_) {
-    layer->Preroll(context, child_matrix);
+    layer->Preroll(context, matrix);
 
     if (layer->needs_system_composite()) {
       set_needs_system_composite(true);
     }
-    child_paint_bounds->join(layer->paint_bounds());
+    child_paint_bounds.join(layer->paint_bounds());
   }
+  set_paint_bounds(child_paint_bounds);
+
+  // Restore the elevation for our parent.
+  context->total_elevation = parent_elevation_;
```

Comment:
Why are we subtracting this entity's elevation from the total here?

Reply:
The context tracks the total elevation in a stack-like fashion as we are walking the LayerTree depth first.  Since we've finished computing elevations for the children of this ContainerLayer, we want to "pop" them off the stack so their elevations won't be incorrectly added into the total for subsequent "branches" of the LayerTree.

For example:

Stack(children: [
    PhysicalShape(elevation: 1.0f, child: PhysicalShape(elevation: 1.0f)),
    PhysicalShape(elevation: 1.0f)
]

would be incorrect without this line of code.  The 2nd PhysicalShape in the Stack would end up with an elevation of 3, instead of 1.

Reply:
Got, thank you! Makes sense. May be worth adding this explanation as a comment inline.